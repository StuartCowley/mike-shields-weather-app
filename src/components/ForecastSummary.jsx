import React from "react";
import PropTypes from "prop-types";
import WeatherIcon from "react-icons-weather";
import ordinalDate from "../helpers/ordinalDate";

export default function ForecastSummary({
  date,
  description,
  icon,
  temperature,
}) {
  const dateString = new Date(date).toDateString();
  const [weekday, month, dayOfMonth] = dateString.split(" ");
  const formattedDate = `${weekday} ${ordinalDate(+dayOfMonth)} ${month}`;

  return (
    <div className="forecast-summary" data-testid="forecast-summary">
      <h2 className="forecast-summary__date">{formattedDate}</h2>
      <p className="forecast-summary__description">{description}</p>
      <WeatherIcon
        className="forecast-summary__icon"
        data-testid="forecast-summary-icon"
        iconId={icon}
        name="owm"
      />
      <p className="forecast-summary__temperature">{temperature.max}</p>
    </div>
  );
}

ForecastSummary.propTypes = {
  date: PropTypes.number.isRequired,
  description: PropTypes.string.isRequired,
  icon: PropTypes.string.isRequired,
  temperature: PropTypes.shape({
    max: PropTypes.number,
    min: PropTypes.number,
  }).isRequired,
};
