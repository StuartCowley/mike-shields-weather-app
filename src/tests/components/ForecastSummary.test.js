import React from "react";
import { render, fireEvent } from "@testing-library/react";
import ForecastSummary from "../../components/ForecastSummary";
import forecast from "../../data/forecast.json";
import unixTimestampToHumanFormattedDate from "../../helpers/unixTimestampToFormattedDate";

const dailyForecast = forecast.forecasts[0];
const { date, description, temperature } = dailyForecast;
const formattedDate = unixTimestampToHumanFormattedDate(date);

describe("ForecastSummary", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(
      <ForecastSummary handleForecastSelect={jest.fn()} {...dailyForecast} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders formatted date, description, icon and temperature", () => {
    const { getByText, getByTestId } = render(
      <ForecastSummary
        data-testid="forecast-summary"
        handleForecastSelect={jest.fn()}
        {...dailyForecast}
      />
    );

    expect(getByText(formattedDate)).toBeInTheDocument();
    expect(getByText(description)).toBeInTheDocument();
    expect(getByTestId("weather-icon-component")).toBeInTheDocument();
    expect(getByText(temperature.max)).toBeInTheDocument();
  });

  it("More details... button should call handleForecastSelect callback", () => {
    const handleForecastSelect = jest.fn();

    const { getByText } = render(
      <ForecastSummary
        data-testid="forecast-summary"
        handleForecastSelect={handleForecastSelect}
        {...dailyForecast}
      />
    );

    const detailsButton = getByText("More details...");

    fireEvent.click(detailsButton);

    expect(handleForecastSelect).toHaveBeenCalled();
  });
});
