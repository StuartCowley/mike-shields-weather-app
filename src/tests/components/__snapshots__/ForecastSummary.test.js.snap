// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastSummary matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="forecast-summary"
    data-testid="forecast-summary"
  >
    <h2
      class="forecast-summary__date"
      data-testid="forecast-summary-date"
    >
      <span>
        Mon 30th Apr
      </span>
    </h2>
    <p
      class="forecast-summary__description"
      data-testid="forecast-summary-description"
    >
      Clear
    </p>
    <div
      class="weather-icon-component"
      data-testid="weather-icon-component"
    >
      <i
        class="wi wi-day-sunny"
        name="owm"
      />
    </div>
    <p
      class="forecast-summary__temperature"
      data-testid="forecast-summary-temperature"
    >
      <span>
        11
      </span>
      Â°C
    </p>
    <button
      class="shadow"
      type="button"
    >
      More details...
    </button>
  </div>
</DocumentFragment>
`;
